File: crates/grida-canvas/src/surface/event.rs
use math2::vector2::Vector2;

/// Modifier keys held during an input event.
#[derive(Debug, Clone, Copy, Default)]
pub struct Modifiers {
    pub shift: bool,
    pub alt: bool,
    pub ctrl_or_cmd: bool,
}

/// Which pointer button was pressed or released.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PointerButton {
    Primary,
    Secondary,
    Middle,
}

/// IME composition event.
#[derive(Debug, Clone)]
pub enum ImeEvent {
    /// Active composition string (displayed inline with underline).
    Preedit(String),
    /// Composition committed — insert the final text.
    Commit(String),
    /// Composition cancelled — discard preedit.
    Cancel,
}

/// Platform-agnostic input events for the canvas surface.
///
/// Coordinates are provided in both canvas space (world) and screen space
/// (viewport pixels). The host is responsible for the camera transform
/// before constructing these events.
#[derive(Debug, Clone)]
pub enum SurfaceEvent {
    PointerMove {
        canvas_point: Vector2,
        screen_point: Vector2,
    },
    PointerDown {
        canvas_point: Vector2,
        screen_point: Vector2,
        button: PointerButton,
        modifiers: Modifiers,
    },
    PointerUp {
        canvas_point: Vector2,
        screen_point: Vector2,
        button: PointerButton,
        modifiers: Modifiers,
    },
    ModifiersChanged(Modifiers),

    /// A keyboard key was pressed.
    ///
    /// Uses [`KeyName`](crate::text_edit::session::KeyName) as the
    /// platform-agnostic key identifier.
    KeyDown {
        key: crate::text_edit::session::KeyName,
        modifiers: Modifiers,
    },

    /// Text input from the OS (after IME composition, dead-key resolution, etc.).
    ///
    /// This is the *committed* text that should be inserted. It may arrive
    /// alongside or instead of a `KeyDown` event, depending on the platform.
    TextInput {
        text: String,
    },

    /// IME composition event.
    Ime(ImeEvent),
}

File: packages/grida-canvas-hud/event/event.ts
import type cmath from "@grida/cmath";

/** Modifier-key snapshot at the moment an event was produced. */
export interface Modifiers {
  shift: boolean;
  alt: boolean;
  meta: boolean;
  ctrl: boolean;
}

export const NO_MODS: Modifiers = {
  shift: false,
  alt: false,
  meta: false,
  ctrl: false,
};

export type PointerButton = "primary" | "secondary" | "middle";

/**
 * Input event consumed by `Surface.dispatch`.
 *
 * All coordinates are **screen-space CSS pixels relative to the canvas**.
 * The surface owns the camera and converts to document-space internally.
 */
export type SurfaceEvent =
  | {
      kind: "pointer_move";
      x: number;
      y: number;
      mods: Modifiers;
    }
  | {
      kind: "pointer_down";
      x: number;
      y: number;
      button: PointerButton;
      mods: Modifiers;
    }
  | {
      kind: "pointer_up";
      x: number;
      y: number;
      button: PointerButton;
      mods: Modifiers;
    }
  | {
      kind: "modifiers";
      mods: Modifiers;
    }
  | {
      kind: "wheel";
      x: number;
      y: number;
      dx: number;
      dy: number;
      mods: Modifiers;
    }
  | {
      kind: "key";
      phase: "down" | "up";
      code: string;
      mods: Modifiers;
    }
  | { kind: "blur" };

/** Result of a `Surface.dispatch` call. */
export interface SurfaceResponse {
  needsRedraw: boolean;
  cursorChanged: boolean;
  hoverChanged: boolean;
}

export function emptyResponse(): SurfaceResponse {
  return {
    needsRedraw: false,
    cursorChanged: false,
    hoverChanged: false,
  };
}

/** Re-export Vector2 for convenience. */
export type Vector2 = cmath.Vector2;
